refactor: ios native module for apn react native sample app

ami-aman opened the pull request:
Related to https://github.com/customerio/issues/issues/8965

Includes:
- ios folder with podfile and other required files
- covers ios native module for the app
-

levibostian commented on Apps/APN/.gitignore:
```diff
@@ -0,0 +1,30 @@
+*.iml
+
+# Google-services.json 
+android/app/google-services.json
```
```suggestion
```

In the Android SDK sample apps, we are not ignoring this file. So I suggest we follow the same pattern in this repository.

levibostian commented on Apps/APN/.gitignore:
```diff
@@ -0,0 +1,30 @@
+app_store_connect_creds.json
+ami_app_ci_server-google_cloud_service_account.json
+gc_keys.json
+GoogleService-Info.plist
+node_modules
+.DS_Store
+env.js
+fastlane/report.xml
```
```suggestion
node_modules
.DS_Store
env.js
fastlane/report.xml
```

Some of these are outdated and can be removed since we don't use the files anymore. 

As for the GoogleService-Info.plist, in the iOS SDK sample apps, we are not ignoring this file. So I suggest we follow the same pattern in this repository.

levibostian commented on Apps/APN/.gitignore:
```diff
@@ -0,0 +1,30 @@
+build/
+
+# CocoaPods
+/ios/Pods/
```
I suggest that we move this `.gitignore` file into `Apps/` directory so it applies to all of our sample apps, not just for this 1 sample app. 

Thanks for adding it!

ami-aman replied:
Moved to `Apps/` in [this commit](https://github.com/customerio/customerio-reactnative/pull/136/commits/7d0fc164ba11d16dcb4424e8da83bc80cf531055)

levibostian commented on Apps/APN/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
This file looks like it was copied from the Ami app. I would delete it from this PR. The whole code base has a LICENSE file in the root directory already.

ami-aman replied:
> I would delete it from this PR

Did you delete the file ? I still see the file in this folder. Want to confirm before making any changes.

levibostian replied:
No, this file still exists in branch `aman/apn-ios` and has not yet been deleted.

levibostian commented on Apps/APN/ios/.gitignore:
```diff
@@ -0,0 +1,97 @@
+test-report.*
```
The contents of this file look good. I suggest that we move all of the contents of this file into `Apps/.gitignore` so all sample apps can use these rules.

ami-aman replied:
Moved in [this commit](https://github.com/customerio/customerio-reactnative/pull/136/commits/b38d4031ed106b030e5029cac8a2d07ed25ac044)

levibostian commented on Apps/APN/ios/AppDelegate.h:
```diff
@@ -0,0 +1,10 @@
+#import <UserNotifications/UNUserNotificationCenter.h>
+#import <UserNotifications/UserNotifications.h>
+
+@interface AppDelegate : UIResponder <UIApplicationDelegate, RCTBridgeDelegate, UNUserNotificationCenterDelegate>
```
Is it common for RN apps to use objective-c AppDelegate files?

ami-aman replied:
Yes, react native apps commonly use ObjC code base.

levibostian commented on Apps/APN/ios/AppDelegate.mm:
```diff
@@ -0,0 +1,194 @@
+// Required for the register event.
+- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
+{
+  [pnHandlerObj application:application deviceToken:deviceToken];
```
I think this function is missing a `super` call. This app might not require it, but anyone who reads this sample app source code and uses it as reference would benefit from adding `super` to their apps.

levibostian commented on Apps/APN/ios/AppDelegate.mm:
```diff
@@ -0,0 +1,194 @@
+- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo
+fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler
+{
+  [RNCPushNotificationIOS didReceiveRemoteNotification:userInfo fetchCompletionHandler:completionHandler];
```
missing `super`

levibostian commented on Apps/APN/ios/AppDelegate.mm:
```diff
@@ -0,0 +1,194 @@
+// Required for the registrationError event.
+- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
+{
+  [pnHandlerObj application:application error:error];
```
missing `super`

levibostian commented on Apps/APN/ios/AppDelegate.mm:
```diff
@@ -0,0 +1,194 @@
+didReceiveNotificationResponse:(UNNotificationResponse *)response
+         withCompletionHandler:(void (^)(void))completionHandler
+{
+  [pnHandlerObj userNotificationCenter:center didReceiveNotificationResponse:response withCompletionHandler:completionHandler];
```
missing `super`? I think? I am not sure if `UserNotification` functions usually call super.

ami-aman replied:
`UserNotification` does not have a `super` call.

levibostian commented on Apps/APN/ios/MyAppPushNotificationsHandler.swift:
```diff
@@ -0,0 +1,47 @@
+  public override init() {}
+
+  // Skip this function if you don't want to request push permissions on launch
+  @objc(registerPushNotification:)
+  public func registerPushNotification(withNotificationDelegate notificationDelegate: UNUserNotificationCenterDelegate) {
+
+    let center  = UNUserNotificationCenter.current()
+    center.delegate = notificationDelegate
+    center.requestAuthorization(options: [.sound, .alert, .badge]) { (granted, error) in
+      if error == nil{
+        DispatchQueue.main.async {
+          UIApplication.shared.registerForRemoteNotifications()
+        }
+      }
+    }
+  }
```
I strongly suggest that our RN sample apps use the SDK's request push permission feature instead of doing this. Since that's our recommended way of requesting push permissions for customers, our sample apps should use the recommended approaches.

ami-aman replied:
Oops! My bad. Not using this function anywhere. Instead calling `showPromptForPushNotifications` to show push permission prompt. Thanks for highlighting this.

ami-aman replied:
Removed it in [this commit](https://github.com/customerio/customerio-reactnative/pull/136/commits/999139686ef2c79fe835f79fb0a1c3e3cdef8e16).

levibostian commented on Apps/APN/ios/Podfile:
```diff
@@ -0,0 +1,43 @@
+  # Flags change depending on the env values.
+  flags = get_default_flags()
+
+  pod 'CustomerIO/MessagingPushAPN', '2.1.2'
```
```suggestion
  pod 'CustomerIO/MessagingPushAPN', '~> 2'
```

Our docs suggest this syntax.

levibostian replied:
I suggest that we add [these functions](https://github.com/customerio/amiapp-reactnative/blob/d1f74c6478fc1b9cf4eb49b2b500cf997a5f0ef8/ios/Podfile#L20-L23) to this file to make development easier for future mobile engineers who want to install pre-production versions of the native iOS SDK into the sample app.

ami-aman replied:
Great suggestion. I would like to create another PR for this change to avoid conflicts in other PRs as this branch is a base to others.

ami-aman replied:
> Our docs suggest this syntax.

Updated this another commit as syntax update was also required for NSE.

levibostian commented on Apps/APN/ios/Podfile:
```diff
@@ -0,0 +1,43 @@
+end
+
+target 'NotificationServiceExtension' do
+  pod 'CustomerIO/MessagingPushAPN', '2.1.2'
```
```suggestion
  pod 'CustomerIO/MessagingPushAPN', '~> 2'
```

ami-aman replied:
I see this now lol. Have already made fix in [this commit.](https://github.com/customerio/customerio-reactnative/pull/136/commits/008d7bfb040cace131b77a9c8516bd64fbdc5831)

levibostian commented on Apps/APN/ios/Podfile:
```diff
@@ -0,0 +1,43 @@
+  pod 'CustomerIO/MessagingPushAPN', '2.1.2'
+end
+
+post_install do |installer|
+  installer.pods_project.targets.each do |target|
+    if target.respond_to?(:product_type) and target.product_type == "com.apple.product-type.bundle"
+      target.build_configurations.each do |config|
+          config.build_settings['CODE_SIGNING_ALLOWED'] = 'NO'
+      end
+    end
+  end
+end
```
What is this for?

ami-aman replied:
I recall adding it as a solution to address code signing errors I encountered with certain libraries. But, I cannot recall specifically which libraries were causing the issue. 🤔

ami-aman replied:
I have removed this snippet as I did not encounter any error but it could be XCode/React native version specific. I will add it back with a comment if required. I hope that will help.

ami-aman commented:
> Thanks for responding to my suggestions, @ami-aman. I see some suggestions have not been responded to. GitHub is hiding them for me so maybe they are being hidden for you, too?
> 
> <img alt="CleanShot 2023-05-22 at 09 56 16@2x" width="1173" src="https://user-images.githubusercontent.com/2041082/239946836-e9f28277-731f-4058-ba0e-b2214f427e57.png">

Thanks Levi. I have addressed all suggestions on this PR now.
